### PR TITLE
npm: add option to package UPM modules as sources for publishing on NPM

### PR DESCRIPTION
**Pull request by Propanu**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -279,6 +285,14 @@ macro(upm_doxygen)
   endif ()
 endmacro(upm_doxygen)
 
+if (NPM)
+add_custom_target (npmpkg)
```
Comment (pylbert) [inline src/CMakeLists.txt]: Should be able to merge the add_custom_command into this add_custom_target (w/o the POST_BUILD).

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -291,7 +305,57 @@ if (BUILDSWIGNODE)
...
   )
   macro(createpackagejson)
+    if (NPM)
+      # Dependency to generate wrappers before packing, skip modules that depend on 3rd party libs
+      if (NOT DEFINED reqlibname)
+        set (reqlibname "")
+      endif ()
+      if (reqlibname STREQUAL "" OR reqlibname MATCHES "mraa")
+        # Grab all module sources and headers, also add sources to gyp list
+        file (GLOB srcfiles RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}"
```
Comment (pylbert) [inline src/CMakeLists.txt]: It is possible to get the SOURCES property for both the base library and NODE wrapper library targets which would provide a finite list of src/hdr files to use.

File: src/binding.gyp.in [hunk 3]
```diff
@@ -0,0 +1,37 @@
+{
```
Comment (pylbert) [inline src/binding.gyp.in]: How close to json does this file need to be?  If this file works then disregard this comment, however you may need to run this by a json lint checker to clean it up a bit.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -291,7 +305,57 @@ if (BUILDSWIGNODE)
...
+        set (upm_LIB_INCLUDE_DIRS_GYP "'upm',\n${upm_LIB_INCLUDE_DIRS_GYP}")
+
+        # Add readme, package.json for NPM and node-gyp config file
+        configure_file (${PROJECT_SOURCE_DIR}/src/binding.gyp.in ${CMAKE_CURRENT_BINARY_DIR}/jsupm_${libname}/binding.gyp @ONLY)
+        configure_file (${PROJECT_SOURCE_DIR}/src/package.json.npm.in ${CMAKE_CURRENT_BINARY_DIR}/jsupm_${libname}/package.json @ONLY)
+        configure_file (${PROJECT_SOURCE_DIR}/doxy/README.npm.md ${CMAKE_CURRENT_BINARY_DIR}/jsupm_${libname}/README.md)
+
+        add_dependencies (npmpkg jsupm_${libname})
+      else ()
+          message ("Skipping jsupm_${libname}, needs ${reqlibname}")
```
Comment (pylbert) [inline src/CMakeLists.txt]: If this message is NOT required to print to stderr consider using message (STATUS "Skipping... )

File: src/binding.gyp.in (hunk 3 above)
Comment (rwaldron) [inline src/binding.gyp.in]: Gyp files aren't real JSON: https://gyp.gsrc.io/docs/UserDocumentation.md

File: src/CMakeLists.txt (hunk 1 above)
Comment (Propanu) [inline src/CMakeLists.txt]: Agreed. Missed it as they were added at different times.

File: src/CMakeLists.txt (hunk 4 above)
Comment (Propanu) [inline src/CMakeLists.txt]: Makes it easier when using a visual interface to cmake (e.g. ccmake), depends on preference. With STATUS they only briefly show on the bottom "status" bar and don't show up in the end report of the configuration step.

**Pull request closed without merging**